### ISSUE TITLE
Condition `step_percentile()` on another feature, add `keep_original_cols` arg

### ISSUE THREAD

**Issue by joeycouse** [tidymodels/recipes#929]
It would be great to have to the ability to condition the use of `step_percentile()` on another feature(s), and have the option to keep the output of the step using `keep_original_cols` which is present in many other step functions.

In the example below, the relationship between time and distance is bimodal, and I would like to filter out the observations that have a much higher distance to time relationship. This creates a much more accurate parameter estimate for the model.

``` r
library(tidymodels)
#> Registered S3 method overwritten by 'tune':
#>   method                   from   
#>   required_pkgs.model_spec parsnip
library(tidyverse)

set.seed(24)

race_data <- tibble(distance = sample(c(2,3,4), size = 1000, replace = T))

race_data <- 
  race_data |> 
  rowwise() |> 
  mutate(final_time = sample(c(rnorm(1, mean = 3, sd = 0.1) * distance,
                             rnorm(1, mean = 7, sd = 0.1) * distance),
                             1,
                             prob = c(0.95,0.05))
         )


# What would be awesome to do inside a reciepe
race_filtered <-
  race_data |> 
  group_by(distance) |> 
  mutate(percentile = ecdf(final_time)(final_time)) |> 
  filter(percentile < 0.90)


# Coefficent Estimate is skewed higher because of outlier points
lm(final_time ~ distance, data = race_data)
#> 
#> Call:
#> lm(formula = final_time ~ distance, data = race_data)
#> 
#> Coefficients:
#> (Intercept)     distance  
#>      -0.394        3.352

# Coefficent Estimate is right on
lm(final_time ~ distance, data = race_filtered)
#> 
#> Call:
#> lm(formula = final_time ~ distance, data = race_filtered)
#> 
#> Coefficients:
#> (Intercept)     distance  
#>    -0.02693      2.99872


# An example of the how the API would work

race_rec <- 
  recipe(final_time ~ distance, data = race_data) |> 
  step_percentile(final_time | distance, keep_original_cols = T) |> 
  step_filter(final_time_percentile < 0.90)
```

<sup>Created on 2022-03-10 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by juliasilge** [comment]
I like this idea. What do you think @topepo? Add a `prefix` argument like `"percentile"` and also `keep_original_cols = FALSE`? This would change the output of old recipes, which is a bummer.

**Comment by brunocarlin** [comment]
I may be able to incorporate this into my tidy.outliers package

**Comment by EmilHvitfeldt** [comment]
I like the idea of the functionality.

I'm not strictly opposed to adding a `keep_original_cols` argument, but I do want to note that it would be a new domain for the argument. Right now it is used for `1-to-many` steps such as `step_date()`, `step_dummy`, and `step_harmonic()` and `many-to-many` steps such as `step_pca()` and `step_pls()`. In my mind I see `step_percentile()` to be the in the same realm of steps as `step_log()`, `step_sqrt()`, and `step_BoxCox()`, and I wouldn't be able to formalize why we would add the argument to this step but not the others. (It is a feature request, which in-and-of itself could be a good enough reason)

Correct me if I'm wrong, but this example would only work for training data right? Since we are filtering according to a derivative of the response we would need to use some `skip = TRUE` in the steps to make this work.

```
race_rec <- 
  recipe(final_time ~ distance, data = race_data) |> 
  step_percentile(final_time | distance, skip = TRUE) |> 
  step_filter(final_time_percentile < 0.90, skip = TRUE)
```

I would advice this functionality to be in [{tidy.outliers}](https://github.com/brunocarlin/tidy.outliers) as I think it would fit in really well. This is in essence a outlier removal technique, and all the steps in that package have `skip = TRUE` as well.

**Comment by brunocarlin** [comment]
Although I like the formula notation, I have very little experience manipulating formulas, I can get this to work quickly if the user passes all the variables used on the step and we assume 'outcome' is the 'final_time' or the user passes the 'final_time' again on this step.

**Comment by joeycouse** [comment]
@brunocarlin formula is not a must for this maybe something like this, with the caveat that the `condition_on` must be factor. 

```r
race_rec <- 
  recipe(final_time ~ distance, data = race_data) |> 
  step_percentile(final_time, condition_on = vars(distance), keep_original_cols = T) |> 
  step_filter(final_time_percentile < 0.90)
```

@EmilHvitfeldt I don't think it would be 100% necessary for `skip = TRUE`, although end user could be surprised when not all rows are returned, but that could be a warning. Code below would achieve what I was thinking although `rowwise()` operations might not be practical on large datasets. 

```r
race_train <-
  race_data |> 
  slice_head(n = 800) 

race_test <-
  race_data |> 
  slice(801:1000) |> 
  rowwise() |> 
  mutate(final_time_percentile = ecdf(race_train[race_train$distance == distance,]$final_time)(final_time))

```

**Comment by brunocarlin** [comment]
I don't like the idea of outlier removal on the testing/validation set, it can cause a lot of problems in implementation, I have ported some packages that do fit a 'model' and can 'predict' on new data, but I would prefer to avoid that if possible.

**Comment by EmilHvitfeldt** [comment]
@joeycouse the code example you posted wouldn't work as `race_test` doesn't have a `final_time`, so it would error when trying to use it at the end of the mutate call `(final_time)`.

**Comment by joeycouse** [comment]
> @joeycouse the code example you posted wouldn't work as `race_test` doesn't have a `final_time` column if applied to new data. It would error when trying to use it at the end of the mutate call `(final_time)`.

Okay gotcha, I missed the most obvious part!

**Comment by brunocarlin** [comment]
Here is the inital implementation I am doing a pr right now on tidy.outliers

``` r
library(tidymodels)
#> Registered S3 method overwritten by 'tune':
#>   method                   from   
#>   required_pkgs.model_spec parsnip
library(tidyverse)
library(tidy.outliers)

set.seed(24)

race_data <- tibble(distance = sample(c(2,3,4), size = 1000, replace = T))

race_data <- 
  race_data |> 
  rowwise() |> 
  mutate(final_time = sample(c(rnorm(1, mean = 3, sd = 0.1) * distance,
                               rnorm(1, mean = 7, sd = 0.1) * distance),
                             1,
                             prob = c(0.95,0.05))
  ) |> 
  ungroup()


race_filtered <-
  race_data |> 
  group_by(distance) |> 
  mutate(percentile = ecdf(final_time)(final_time)) |> 
  filter(percentile < 0.90)

lm(final_time ~ distance, data = race_data)
#> 
#> Call:
#> lm(formula = final_time ~ distance, data = race_data)
#> 
#> Coefficients:
#> (Intercept)     distance  
#>      -0.394        3.352


lm(final_time ~ distance, data = race_filtered)
#> 
#> Call:
#> lm(formula = final_time ~ distance, data = race_filtered)
#> 
#> Coefficients:
#> (Intercept)     distance  
#>    -0.02693      2.99872


rec <- recipe(final_time ~ distance,data = race_data)

juiced <- rec |> 
  step_outliers_ecdf(everything(),name_mutate = 'percentile') |>
  step_outliers_remove('percentile',score_dropout = .9) |> 
  prep(race_data) |> 
  juice()


lm(final_time ~ distance, data = juiced)
#> 
#> Call:
#> lm(formula = final_time ~ distance, data = juiced)
#> 
#> Coefficients:
#> (Intercept)     distance  
#>    -0.02693      2.99872
```

<sup>Created on 2022-03-14 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by EmilHvitfeldt** [comment]
I'm gonna close this issue as the proposed functionality has been implemented in https://github.com/brunocarlin/tidy.outliers
